---
title: "RNA copies per cell"
author: "Amalia MS"
date: "Tuesday, April 19, 2016"
output: html_document
---

After fixing my stupid mistake multiplying and using DESeq normalised reads, it seems the copies per cell of RNA molecules looks reasonable. I'm using the linear regression of fitting al the biological replicates at the same time. 

```{r, echo = F, message= F}
rm(list = ls())

library(ggplot2)
library(reshape2)
library(rootSolve)
library(polynom)
library(plyr)
library(DESeq2)
library(gtools)
library(gplots)
library(RColorBrewer)

knitr::opts_knit$set(root.dir = 'C:/Users/am4613/Documents/Summaries_as_timecourses/')

```

#### Scatterplot comparing my data with the one published in Marguerat et al. 2012. Why is it so weird?

```{r, echo=FALSE}
#Create table with all the data

rna_cpc <- read.delim('rna_copies_per_cell.txt', header = T, strings = F)

sam_cpc <- read.delim('sams_data.txt', skip = 6, header = T, strings = F)

all_data <- merge(rna_cpc, sam_cpc, by.x = 'row.names', by.y = 'Systematic.name', all.x = T)
plot(y = log2(all_data$rev_tp0_sho1.bam), x = log2(all_data$MM.mRNA.cpc),
		 xlab = 'Log2(Sam data)', ylab = 'Log2(Amalia data)')
abline(0,1, col = 'darkgrey')


```

#####Is it weird because of the calibration or because of the rpkms themselves?

```{r, echo = FALSE}

#Load counts and normalise using DESeq2
all_counts <- read.delim('analysis/all_rev_counts.txt', header = T, strings = F)
row.names(all_counts) <- all_counts$ID
all_counts <- all_counts[,2:ncol(all_counts)]
all_counts <- all_counts[,mixedorder(colnames(all_counts))]

exp_design = data.frame(Time = factor(rep(0:11)), 
												Replicate = rep(1:3,12), 
												group = rep(1,ncol(all_counts)),
												row.names = colnames(all_counts))


cds <- DESeqDataSetFromMatrix(countData = all_counts, colData = exp_design, design = ~Time)
cds <- estimateSizeFactors(cds)
size_fact <- sizeFactors(cds)

rpkm <- all_counts

for(i in 1:36)
{
	rpkm[,i] <- all_counts[,i]/size_fact[i]
}

sam_rpkm <- merge(rpkm, sam_cpc, by.x = 'row.names', by.y = 'Systematic.name', all.x = T)
plot(y = log2(sam_rpkm$rev_tp0_sho1.bam), x = log2(sam_rpkm$MM.mRNA.cpc), xlab = 'Log2(Sam copies per cell)', ylab = 'Log2(Normalised counts tp0 Amalia)', main = 'Repeat1')
abline(0,1,col = 'darkgray')

plot(y = log2(sam_rpkm$rev_tp0_sho2.bam), x = log2(sam_rpkm$MM.mRNA.cpc), xlab = 'Log2(Sam copies per cell)', ylab = 'Log2(Normalised counts tp0 Amalia)', main = 'Repeat2')
abline(0,1,col = 'darkgray')

plot(y = log2(sam_rpkm$rev_tp0_sho3.bam), x = log2(sam_rpkm$MM.mRNA.cpc), xlab = 'Log2(Sam copies per cell)', ylab = 'Log2(Normalised counts tp0 Amalia)', main = 'Repeat3')
abline(0,1,col = 'darkgray')


sam_counts <- read.delim('sam_counts.txt', header = T, strings = F, skip = 1)

merg_counts <- merge(rpkm, sam_counts, by.x = 'row.names', by.y = 'Systematic.name', all.x = T)

plot(y = log2(merg_counts$rev_tp0_sho1.bam), x = log2(merg_counts$MM1.tot.cRPK_ex), xlab = 'Log2(RPKM Sam)', ylab = 'Log2(Normalised counts tp0 Amalia)', main = 'Repeat1')

plot(y = log2(merg_counts$rev_tp0_sho2.bam), x = log2(merg_counts$MM1.tot.cRPK_ex), xlab = 'Log2(RPKM Sam)', ylab = 'Log2(Normalised counts tp0 Amalia)', main = 'Repeat2')

plot(y = log2(merg_counts$rev_tp0_sho3.bam), x = log2(merg_counts$MM1.tot.cRPK_ex), xlab = 'Log2(RPKM Sam)', ylab = 'Log2(Normalised counts tp0 Amalia)', main = 'Repeat3')

```


####There are some guys that seem to be weirdly correlated in the plot before, who are those guys?

```{r, echo = FALSE}


nc_rna <- all_data[grep(all_data$Row.names, pattern = 'SPNCRNA'),]
r_rna <- all_data[grep(all_data$Row.names, pattern = 'SPRRNA'),]
sno_rna <- all_data[grep(all_data$Row.names, pattern = 'SPSNO'),]
t_rna <- all_data[grep(all_data$Row.names, pattern = 'TRNA'),]

plot(y = log2(all_data$rev_tp0_sho1.bam), x = log2(all_data$MM.mRNA.cpc),
		 xlab = 'Log2(Sam data)', ylab = 'Log2(Amalia data)')
points(y = log2(nc_rna$rev_tp0_sho1.bam), x = log2(nc_rna$MM.mRNA.cpc), col = 'red')
points(y = log2(r_rna$rev_tp0_sho1.bam), x = log2(r_rna$MM.mRNA.cpc), col = 'green')
points(y = log2(sno_rna$rev_tp0_sho1.bam), x = log2(sno_rna$MM.mRNA.cpc), col = 'blue')
points(y = log2(t_rna$rev_tp0_sho1.bam), x = log2(t_rna$MM.mRNA.cpc), col = 'purple')

legend('topleft', c('Non coding','rrna','snorna','trna','mRNA'), fill = c('red','green','blue','purple','black'))


```

#### Comparing distributions of both datasets (my data in red). Bear in mind that there are more total mRNA in Sam's dataset (rRNA not represented in our data?)

```{r, echo = F}

plot(density(log2(all_data$MM.mRNA.cpc), na.rm = T), ylim = c(0,0.5))
lines(density(log2(all_data$rev_tp0_sho1.bam)), col = 'red', main = 'Both dataset distribution')

hist(log2(all_data$MM.mRNA.cpc), main = 'Histogram Sam dataset')
hist(log2(all_data$rev_tp0_sho1.bam), main = 'Histogram Me dataset')
```


#### Comparing all time points between each other. It seems the increase is more or less coordinated between time points.

```{r, echo = T}

plot(log2(all_data$rev_tp0_sho1.bam), log2(all_data$rev_tp11_sho1.bam))
abline(0,1,col = 'grey')

data <- log2(all_data[,2:13])

#Melt puts all the columns of the dataframe in one, with the colname as variable
m_data <- melt(data)
ggplot(m_data, aes(x = value, fill = variable)) + geom_density(alpha = 0.5)
```

####Heatmap of the data. Log2 and time point zero normalised

```{r, echo = F}
rna_cpc_forlog <- rna_cpc+0.001
norm_cpc <- rna_cpc_forlog/rna_cpc_forlog[,1]
col = colorRampPalette(brewer.pal(9,'GnBu'))
heatmap.2(as.matrix(log2(norm_cpc)), col = col, Colv = F)

```

####Fit third, second and firt degree polynomials to all the data, choosing the best one, try to find factors whose first or second derivative are different than average.

```{r, echo = T}
time = c(0:11)

data <- all_data[,2:13]

best_model <- list()

##Get the best model for every gene, choosing according to adjusted r squared

for(i in 1:nrow(data))
{
	test <- cbind(t(data[i,]),time)
	colnames(test) <- c('gene','time')
	test <- as.data.frame(test)
	poly1 <- lm(test$gene ~ poly(test$time,1,raw=TRUE))
	poly2 <- lm(test$gene ~ poly(test$time,2,raw=TRUE))
	poly3 <- lm(test$gene ~ poly(test$time,3,raw=TRUE))
	
	adj_r_squared <- c(summary(poly1)$adj.r.squared,summary(poly2)$adj.r.squared,summary(poly3)$adj.r.squared)
	pos <- which(adj_r_squared == max(adj_r_squared))
	
	#There are genes that have models with same r squared, we pick the simplest one
	if(length(pos) > 1)
		{
			pos = min(pos)
		}
	
	
	if(pos == 1)
	{
		best_model[[i]] <- poly1
	}else if(pos == 2)
	{
		best_model[[i]] <- poly2
	}else if (pos == 3)
	{
		best_model[[i]] <- poly3
	}
}

#Get coefficients from model and remove NA
poly_coeff <- lapply(best_model, coefficients)
poly_coeff <- rapply(poly_coeff, f=function(x) ifelse(is.na(x),0,x), how="replace")

#Convert those coefficients to polynomials
poly_model <- lapply(poly_coeff, polynomial)

#First derivative and finding its solutions
first_deriv <- lapply(poly_model, deriv)
first_sol <- lapply(lapply(first_deriv,polyroot),Re)
first_sol <- ldply(first_sol, rbind)
row.names(first_sol) <- all_data$Row.names

hist(melt(first_sol)$value, breaks = 100)
#Second derivative
second_deriv <- lapply(first_deriv, deriv)
second_sol <- lapply(lapply(second_deriv,polyroot),Re)
second_sol <- ldply(second_sol, rbind)
row.names(second_sol) <- all_data$Row.names

hist(melt(second_sol)$value, breaks = 100)

solutions_table <- data.frame(row.names = row.names(first_sol), first_derivative_sol1 = first_sol[,1],
															first_derivative_sol2 = first_sol[,2], second_derivative = second_sol[,1])

